cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -373,6 +374,9 @@ const (
 // NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
 // paired to their InitFunc.  This allows for structured downstream composition and subdivision.
 func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc {
+	if len(controllersMap) > 0 && len(controllersMap[loopMode]) > 0 {
+		return controllersMap[loopMode]
```

Comment:
why is it helpful to return the old map again?

It's a constructor that does not have long runtime. Do we care about running it twice?

It is even dangerous as a `New*` func should return a new object, i.e. the caller has full ownership. Now we return a singleton that causes potential races.

Finally, the constructor is not go-routine safe anymore because `controllersMap` is unprotected.

Reply:
@sttts 
Thanks for providing food for thought.
```
New* func should return a new object
```
The return value from KnownControllers is given to Option / Flag where validation is done.
The known controllers are read only.

Reply:
w.r.t. the last point (concurrency), KnownControllers() is called by NewControllerManagerCommand which returns cobra.Command whose Run func calls KnownControllers() as well.
Since the call to KnownControllers() in Run func follows the completion of NewControllerManagerCommand, I don't see race condition between the two calls.